pkg/controller/deployment_inplace.go

```diff
@@ -0,0 +1,590 @@
+		}
+
+		node, err := dc.nodeLister.Get(nodeName)
+		if err != nil {
+			return fmt.Errorf("failed to get node %s: %w", nodeName, err)
+		}
+
+		cond := nodeops.GetCondition(node, v1alpha1.NodeInPlaceUpdate)
+		if isUpdateNotSuccessful(cond, node.Labels) {
+			continue
+		}
+
+		// remove labels related to the inplace update.
+		nodeLabels := node.Labels
+		delete(nodeLabels, v1alpha1.LabelKeyNodeCandidateForUpdate)
+		delete(nodeLabels, v1alpha1.LabelKeyNodeSelectedForUpdate)
+		delete(nodeLabels, v1alpha1.LabelKeyNodeUpdateResult)
+		// remove annotations related to the inplace update.
+		delete(node.Annotations, v1alpha1.AnnotationKeyMachineUpdateFailedReason)
+
+		node.ObjectMeta.Labels = nodeLabels
+		// uncordon the node since the inplace update is successful.
+		node.Spec.Unschedulable = false
+
+		_, err = dc.targetCoreClient.CoreV1().Nodes().Update(ctx, node, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to uncordon the node %s: %w", node.Name, err)
+		}
+	}
+
+	for _, machineSet := range oldMachineSets {
+		// scale down the old machine set to the number of machines which is having the labelselector of the machine set.
+		oldMachines, err := dc.machineLister.List(labels.SelectorFromSet(machineSet.Spec.Selector.MatchLabels))
+		if err != nil {
+			return fmt.Errorf("failed to list machines for machine set %s: %w", machineSet.Name, err)
+		}
+
+		if len(oldMachines) < int(machineSet.Spec.Replicas) {
+			_, _, err := dc.scaleMachineSetAndRecordEvent(ctx, machineSet, int32(len(oldMachines)), deployment) // #nosec G115 (CWE-190) -- value already validated
+			if err != nil {
+				return fmt.Errorf("failed to scale down machine set %s: %w", machineSet.Name, err)
+			}
+			klog.V(3).Infof("scaled down machine set %s to %d", machineSet.Name, len(oldMachines))
+		}
+	}
+
+	return nil
+}
+
+func (dc *controller) reconcileNewMachineSetInPlace(ctx context.Context, oldMachineSets []*v1alpha1.MachineSet, newMachineSet *v1alpha1.MachineSet, deployment *v1alpha1.MachineDeployment) (bool, error) {
+	if newMachineSet.Spec.Replicas == deployment.Spec.Replicas {
+		// Scaling not required.
+		return false, nil
+	}
+
+	if newMachineSet.Spec.Replicas > deployment.Spec.Replicas {
+		// Scale down.
+		scaled, _, err := dc.scaleMachineSetAndRecordEvent(ctx, newMachineSet, (deployment.Spec.Replicas), deployment)
+		return scaled, err
+	}
+
+	klog.V(3).Infof("reconcile new machine set %s", newMachineSet.Name)
+
+	oldMachinesCount := GetReplicaCountForMachineSets(oldMachineSets)
+	if oldMachinesCount == 0 {
+		// if there are no machines in the old machine sets, then there is no need to transfer the ownership of the machines.
+		// scale up the new machine set to the required replicas.
+		scaleBy := deployment.Spec.Replicas - newMachineSet.Spec.Replicas // here scaleBy is always positive
+		scaled, _, err := dc.scaleMachineSetAndRecordEvent(ctx, newMachineSet, newMachineSet.Spec.Replicas+scaleBy, deployment)
```

Comment:
```suggestion
		scaled, _, err := dc.scaleMachineSetAndRecordEvent(ctx, newMachineSet, deployment.Spec.Replicas, deployment)
```
`scaleBy` isn't required here right, consider removing the variable